src/main/java/org/broadinstitute/hellbender/tools/spark/sv/evidence/FermiLiteAssemblyHandler.java

```diff
@@ -24,51 +26,509 @@
+            newConnections.add(toAdd);
+        }
+        contig.setConnections(newConnections);
+    }
+
+    // join the sequences of a chain of contigs to produce a single, new contig
+    private static Contig joinContigs( final Contig firstContig, final List<Connection> path ) {
+        final int nSupportingReads =
+                path.stream()
+                        .mapToInt(conn -> conn.getTarget().getNSupportingReads())
+                        .reduce(firstContig.getNSupportingReads(), Integer::sum);
+        final int newContigLen =
+                path.stream()
+                        .mapToInt(conn -> conn.getTarget().getSequence().length - conn.getOverlapLen())
+                        .reduce(firstContig.getSequence().length, Integer::sum);
+        final byte[] sequence = new byte[newContigLen];
+        int dstIndex = firstContig.getSequence().length;
+        System.arraycopy(firstContig.getSequence(), 0, sequence, 0, dstIndex);
+        if ( path.get(0).isRC() )
+            SequenceUtil.reverseComplement(sequence, 0, dstIndex);
+        for ( final Connection conn : path ) {
+            final byte[] contigSequence = conn.getTarget().getSequence();
+            final int len = contigSequence.length - conn.getOverlapLen();
+            if ( !conn.isTargetRC() ) {
+                System.arraycopy(contigSequence, conn.getOverlapLen(), sequence, dstIndex, len);
+            } else {
+                System.arraycopy(contigSequence, 0, sequence, dstIndex, len);
+                SequenceUtil.reverseComplement(sequence, dstIndex, len);
+            }
+            dstIndex += len;
+        }
+        return new Contig(sequence, null, nSupportingReads);
+    }
+
+    // walk the graph to trace out all paths, breaking only at cycles and at contigs that require phasing
+    @VisibleForTesting
+    static FermiLiteAssembly expandAssemblyGraph( final FermiLiteAssembly assembly ) {
+        final int nContigs = assembly.getNContigs();
+        // this will hold the new compound contigs that we build
+        final List<Contig> contigList = new ArrayList<>(nContigs);
+
+        // this will keep track of where we've been during a depth-first exploration from some starting contig.
+        // we'll use it to kill cycles.
+        final Set<ContigStrand> visited = new HashSet<>();
+
+        // this will keep track of contigs that we've touched during any of our explorations.
+        // we'll use it to avoid backward tracing from contigs that have already appeared on some path.
+        final Set<Contig> examined = new HashSet<>(SVUtils.hashMapCapacity(nContigs));
+
+        // trace paths from sources and sinks
+        assembly.getContigs().forEach(tig -> {
+            if ( examined.contains(tig) ) return;
+            if ( tig.getConnections().isEmpty() ) {
+                contigList.add(tig);
+                examined.add(tig);
+            } else {
+                final int nPredecessors = countPredecessors(tig);
```

Comment:
Is my understanding correct, that after the call to `removeUnbranchedConnections()`, there could only be 
* isolated islands, 
* cycles, and 
* parallel paths (i.e. bubble, if not an abuse of terms); 

and this block is to extend the parallel paths as much as possible?

Reply:
The graph structure will be arbitrarily complex with isolated islands, and multiple components consisting of cycles and branching structures potentially a lot more complex than implied by "bubble".  But, yeah.  I think you've got the idea.